RFCs/FS-1135-random-functions-for-collections.md

```diff
@@ -0,0 +1,175 @@
+# F# RFC FS-1135 - Random functions for collections (List, Array, Seq)
+
+The design suggestion [Add shuffle, sample etc. methods for lists, arrays etc](https://github.com/fsharp/fslang-suggestions/issues/508) has been marked "approved in principle".
+
+This RFC covers the detailed proposal for this suggestion.
+
+- [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/508)
+- [x] Approved in principle
+- [ ] [Implementation]() (no implementation yet)
+- [ ] Design Review Meeting(s) with @dsyme and others invitees
+- [Discussion](https://github.com/fsharp/fslang-design/discussions/731)
+
+# Summary
+
+This feature extends collections apis with functions for random sampling and shuffling built-in fsharp collections.
+
+# Motivation
+
+This feature is motivated by the following use cases:
+ - Using F# for data science and machine learning (like building a neural network), where data shuffling plays important role
+ - Building games, where random sampling is used for generating random levels, random decks, etc.
+ - Building simulations, where random sampling is used for generating random input data
+
+# Detailed design
+
+### General
+
+The following general rules are applied to all functions
+ - New function should be implemented in List, Array, Seq modules
+ - All functions should not mutate the input collection
+ - All functions should have a variant with a [Random](https://learn.microsoft.com/en-us/dotnet/api/system.random) parameter
+ - Shared thread-safe Random instance should be used for all basic functions.
+
+### Shuffle
+
+Shuffle function should returned a new shuffled collection of the same collection type.
+
+Two functions should be added to each module.
+
+```fsharp
+val shuffle: CollectionType<'T> -> CollectionType<'T>
+val shuffleRand: Random -> CollectionType<'T> -> CollectionType<'T>
```

Comment:
What is CollectionType?

As per the issue these are supposed to sit on the major F# collection types. I'd like to see a proposed signature for each.

Reply:
Fixed